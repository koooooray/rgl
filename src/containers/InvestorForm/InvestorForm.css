.InvestorForm {
    margin: 30px auto;
    width: 75%;
    max-width: 800px;
    min-width: 600px;
    padding: 30px;
    box-sizing: border-box;
    border: 1px seashell solid;
    background-color: beige;
    display: flex;
    flex-direction: row;
}

.InvestorForm form{
    width: 100%;
}