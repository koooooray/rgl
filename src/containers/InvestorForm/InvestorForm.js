import React, { Component } from "react";
import classes from "./InvestorForm.css";
import InputElement from "../../components/InputElement/InputElement";
import Button from "../../components/Button/Button";

class InvestorForm extends Component {
  countries = [
    { value: "usa",  supportedCountry: true, displayName: "United States of America", selected: false },
    { value: "ger", supportedCountry: true, displayName: "Germany", selected: false },
    { value: "fra", supportedCountry: false, displayName: "France", selected: false },
    { value: "uk", supportedCountry: true, displayName: "United Kingdom", selected: false },
    { value: "hol", supportedCountry: false, displayName: "Netherlands", selected: false },
    { value: "can", supportedCountry: false, displayName: "Canada", selected: false }
  ];
  
  state = {
    investorForm: {
      country: {
        elementType: "select",
        elementConfig: {
          options: this.countries,
          hasEmptyOption: true
        },
        value: "",
        label: "Country of citizenship",
        touched: false,
        required: true,
        valid: false,
        validationMessage: "initial",
        checkValidity: element => {
          const country = this.countries.find(item => item.value === element.value);
          const result = element.value === "usa" || (country && country.supportedCountry);
          if(!result){
            if(!country){
              element.validationMessage = `Ineligible: Please select your country.`;
            }else{
              element.validationMessage = `Ineligible: We donâ€™t currently support investors in ${country.displayName}.`;
            }
          } else{
            element.validationMessage = "";
          }
          return result;
        }
      },
      accredited: {
        elementType: "toggle",
        elementConfig: {
          checked:false
        },
        value: false,
        label: "Are you an accredited investor?",
        touched: false,
        required: true,
        valid: false,
        validationMessage: "initial",
        checkValidity: element => {
          if(!element.value){
            element.validationMessage = "Ineligible: We are only able to accept accredited investors at this time";
          } else {
            element.validationMessage = "";
          }
          return element.value;
        }
      },
      identityProof: {
        elementType: "toggle",
        elementConfig: {
          checked: false
        },
        value: false,
        label: "Can you provide proof of identity?",
        touched: false,
        required: true,
        valid: false,
        validationMessage: "initial",
        checkValidity: element => {
          if(!element.value){
            element.validationMessage = "Ineligible: We must be able to verify your identity";
          } else {
            element.validationMessage = "";
          }
          return element.value;
        }
      }
    },
    isFormValid: false,
    isLoading: false
  };

  validateInputConfiguration(inputConfiguration) {
    return inputConfiguration.elementType 
      && inputConfiguration.label 
      && inputConfiguration.checkValidity && typeof inputConfiguration.checkValidity === 'function';
  }

  onInputChanged = (event, inputId) => {

    const updatedInvestorForm = { ...this.state.investorForm };
    const updatedFormElement = { ...this.state.investorForm[inputId] };

    updatedFormElement.value =
      updatedFormElement.elementType === "toggle"
        ? event.target.checked
        : event.target.value;
        
    updatedFormElement.valid = updatedFormElement.checkValidity(updatedFormElement);
    updatedFormElement.touched = true;

    updatedInvestorForm[inputId] = updatedFormElement;

    let formIsValid = true;
    for (const key in updatedInvestorForm) {
      formIsValid = updatedInvestorForm[key].valid && updatedInvestorForm[key].touched && formIsValid;
    }

    this.setState({ investorForm: updatedInvestorForm, isFormValid: formIsValid });
  };

  render() {
    const formElements = [];
    for (const key in this.state.investorForm) {
      const inputConfiguration = this.state.investorForm[key];
      if(this.validateInputConfiguration(inputConfiguration)){
        if(!formElements.find(item=>item.key === key)){
          formElements.push({
            id: key,
            configuration: inputConfiguration
          });
        }
      }
    }

    let formToBeRendered = (
      <form>
        {formElements.map(formElement => (
          <InputElement
            key={formElement.id}
            elementType={formElement.configuration.elementType}
            label={formElement.configuration.label}
            value={formElement.configuration.value}
            elementConfig={formElement.configuration.elementConfig}
            invalid={!formElement.configuration.valid}
            validationMessage={formElement.configuration.validationMessage}
            touched={formElement.configuration.touched}
            changed={event => this.onInputChanged(event, formElement.id)}
          />
        ))}
        <Button btnType="Success" disabled={!this.state.isFormValid}>
          VERIFY
        </Button>
      </form>
    );

    if (this.state.isLoading) {
      formToBeRendered = <p>Loading form...</p>;
    }

    return <div className={classes.InvestorForm}>{formToBeRendered}</div>;
  }
}

export default InvestorForm;
