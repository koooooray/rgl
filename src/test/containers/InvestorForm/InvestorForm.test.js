import InvestorForm from "../../../containers/InvestorForm/InvestorForm";
import React from "react";
import ReactShallowRenderer from "react-test-renderer/shallow";

import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });
import { render, mount, shallow } from "enzyme";

test("should skip empty input objects", () => {
  const invForm = new InvestorForm();
  const input = {};
  expect(invForm.validateInputConfiguration(input)).toBeFalsy();
});

test("should skip input objects with no type", () => {
  const invForm = new InvestorForm();
  const input = {
    elementType: ""
  };
  expect(invForm.validateInputConfiguration(input)).toBeFalsy();
});

test("should skip input objects with no label", () => {
  const invForm = new InvestorForm();
  const input = {
    elementType: ""
  };
  expect(invForm.validateInputConfiguration(input)).toBeFalsy();
});

test("should skip input objects no checkValidity function", () => {
  const invForm = new InvestorForm();
  const input = {
    elementType: "toggle",
    label: "some-input"
  };
  expect(invForm.validateInputConfiguration(input)).toBeFalsy();
});

test("should validate correctly configured input elements", () => {
  const invForm = new InvestorForm();
  const input = {
    elementType: "toggle",
    label: "some-input",
    checkValidity: () => true
  };
  expect(invForm.validateInputConfiguration(input)).toBeTruthy();
});

test("should render form with tree input fields", () => {
  const renderer = new ReactShallowRenderer();
  renderer.render(<InvestorForm />);
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});

// end to end tests
test("should render one select element and two toggle elements", () => {
  const wrapper = render(<InvestorForm />);
  expect(wrapper.find("select").length).toBe(1);
  expect(wrapper.find("select").get(0).children.length).toBe(7);
  expect(wrapper.find("input").length).toBe(2);
});

test("verify button is disabled", () => {
  const wrapper = render(<InvestorForm />);
  const attr = wrapper.find("button").get(0).attribs;
  expect("disabled" in attr).toBe(true);
});

test("initially there are no validation errors and verify button is disabled", () => {
    const wrapper = mount(<InvestorForm/>);
    expect(wrapper.find(".validation").at(0).text()).toBe('');
    expect(wrapper.find(".validation").at(1).text()).toBe('');
    expect(wrapper.find(".validation").at(2).text()).toBe('');

    const button = wrapper.find("button").at(0);
    expect(button.props().disabled).toBe(true);

    wrapper.find("select").simulate('change', {target: {value: 'ger'}});
    expect(wrapper.find(".validation").at(0).text()).toBe('');
});

test("usa, germany uk are supported countries. Validation message is shown otherwise", () => {
    const wrapper = mount(<InvestorForm/>);
    wrapper.find("select").simulate('change', {target: {value: 'can'}});
    expect(wrapper.find(".validation").at(0).text()).toBe('Ineligible: We don’t currently support investors in Canada.');
    wrapper.find("select").simulate('change', {target: {value: 'usa'}});
    expect(wrapper.find(".validation").at(0).text()).toBe('');
    wrapper.find("select").simulate('change', {target: {value: 'fra'}});
    expect(wrapper.find(".validation").at(0).text()).toBe('Ineligible: We don’t currently support investors in France.');

    const button = wrapper.find("button").at(0);
    expect(button.props().disabled).toBe(true);

    wrapper.find("select").simulate('change', {target: {value: 'ger'}});
    expect(wrapper.find(".validation").at(0).text()).toBe('');
});

test("accredited investors validation message is shown if checkbox is unchecked", () => {
    const wrapper = mount(<InvestorForm/>);
    wrapper.find("input").at(0).simulate('change', {target: {checked: false}});
    expect(wrapper.find(".validation").at(1).text()).toBe('Ineligible: We are only able to accept accredited investors at this time');
    
    const button = wrapper.find("button").at(0);
    expect(button.props().disabled).toBe(true);

    wrapper.find("input").at(0).simulate('change', {target: {checked: true}});
    expect(wrapper.find(".validation").at(1).text()).toBe('');
});

test("ineligible: We must be able to verify your identity", () => {
    const wrapper = mount(<InvestorForm/>);
    wrapper.find("input").at(1).simulate('change', {target: {checked: false}});
    expect(wrapper.find(".validation").at(2).text()).toBe('Ineligible: We must be able to verify your identity');
    
    const button = wrapper.find("button").at(0);
    expect(button.props().disabled).toBe(true);

    wrapper.find("input").at(1).simulate('change', {target: {checked: true}});
    expect(wrapper.find(".validation").at(2).text()).toBe('');
}); 

test("all inputs must have valid values.", () => {
    const wrapper = mount(<InvestorForm/>);
    wrapper.find("select").simulate('change', {target: {value: 'usa'}});
    wrapper.find("input").at(0).simulate('change', {target: {checked: false}});
    wrapper.find("input").at(1).simulate('change', {target: {checked: true}});

    const button = wrapper.find("button").at(0);
    expect(button.props().disabled).toBe(true);

    wrapper.find("select").simulate('change', {target: {value: 'usa'}});
    wrapper.find("input").at(0).simulate('change', {target: {checked: true}});
    wrapper.find("input").at(1).simulate('change', {target: {checked: false}});

    expect(button.props().disabled).toBe(true);

    wrapper.find("select").simulate('change', {target: {value: 'can'}});
    wrapper.find("input").at(0).simulate('change', {target: {checked: true}});
    wrapper.find("input").at(1).simulate('change', {target: {checked: true}});

    expect(button.props().disabled).toBe(true);
}); 

test("with valid inputs verify button is enabled.", () => {
    const wrapper = mount(<InvestorForm/>);
    wrapper.find("select").simulate('change', {target: {value: 'usa'}});
    wrapper.find("input").at(0).simulate('change', {target: {checked: true}});
    wrapper.find("input").at(1).simulate('change', {target: {checked: true}});

    const button = wrapper.find("button").at(0);
    expect(button.props().disabled).toBe(false);
}); 