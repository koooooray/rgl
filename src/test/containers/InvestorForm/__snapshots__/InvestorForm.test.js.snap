// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render form with tree input fields 1`] = `
<div
  className={undefined}
>
  <form>
    <InputElement
      changed={[Function]}
      elementConfig={
        Object {
          "hasEmptyOption": true,
          "options": Array [
            Object {
              "displayName": "United States of America",
              "selected": false,
              "supportedCountry": true,
              "value": "usa",
            },
            Object {
              "displayName": "Germany",
              "selected": false,
              "supportedCountry": true,
              "value": "ger",
            },
            Object {
              "displayName": "France",
              "selected": false,
              "supportedCountry": false,
              "value": "fra",
            },
            Object {
              "displayName": "United Kingdom",
              "selected": false,
              "supportedCountry": true,
              "value": "uk",
            },
            Object {
              "displayName": "Netherlands",
              "selected": false,
              "supportedCountry": false,
              "value": "hol",
            },
            Object {
              "displayName": "Canada",
              "selected": false,
              "supportedCountry": false,
              "value": "can",
            },
          ],
        }
      }
      elementType="select"
      invalid={true}
      label="Country of citizenship"
      touched={false}
      validationMessage="initial"
      value=""
    />
    <InputElement
      changed={[Function]}
      elementConfig={
        Object {
          "checked": false,
        }
      }
      elementType="toggle"
      invalid={true}
      label="Are you an accredited investor?"
      touched={false}
      validationMessage="initial"
      value={false}
    />
    <InputElement
      changed={[Function]}
      elementConfig={
        Object {
          "checked": false,
        }
      }
      elementType="toggle"
      invalid={true}
      label="Can you provide proof of identity?"
      touched={false}
      validationMessage="initial"
      value={false}
    />
    <button
      btnType="Success"
      disabled={true}
    >
      VERIFY
    </button>
  </form>
</div>
`;
